src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,513 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/BasicConfig.h>
+#include <ripple/basics/impl/PerfLogImp.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/json/json_writer.h>
+#include <ripple/json/to_string.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <atomic>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+namespace ripple {
+namespace perf {
+
+void
+PerfLogImp::Counters::populateRpc(std::vector<char const*> const& labels)
+{
+    for ( auto label : labels)
+    {
+        auto const &elem = rpc_.emplace(label, Rpc(label));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+void
+PerfLogImp::Counters::populateJq(JobTypes const& jobTypes)
+{
+    for (auto const& job : jobTypes)
+    {
+        auto elem = jq_.emplace(job.first, Jq(job.second.name()));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+Json::Value
+PerfLogImp::Counters::countersJson() const
+{
+    Json::Value rpcobj(Json::objectValue);
+    // totalRpc represents all rpc methods. All that started, finished, etc.
+    Rpc totalRpc("total");
+    for (auto const& proc : rpc_)
+    {
+        Json::Value p(Json::objectValue);
+        {
+            std::unique_lock<std::mutex> lock(proc.second.mut);
+            auto const tmp = proc.second.sync;
+            lock.unlock();
+            p[jss::started] = std::to_string(tmp.started);
+            totalRpc.sync.started += tmp.started;
+            p[jss::finished] = std::to_string(tmp.finished);
+            totalRpc.sync.finished += tmp.finished;
+            p[jss::errored] = std::to_string(tmp.errored);
+            totalRpc.sync.errored += tmp.errored;
+            p[jss::duration_us] = std::to_string(tmp.duration.count());
+            totalRpc.sync.duration += tmp.duration;
+        }
+        rpcobj[proc.second.label] = p;
+    }
+    Json::Value totalRpcJson(Json::objectValue);
+    totalRpcJson[jss::started] = std::to_string(
+        totalRpc.sync.started);
+    totalRpcJson[jss::finished] = std::to_string(
+        totalRpc.sync.finished);
+    totalRpcJson[jss::errored] = std::to_string(
+        totalRpc.sync.errored);
+    totalRpcJson[jss::duration_us] = std::to_string(
+        totalRpc.sync.duration.count());
+    rpcobj[jss::total] = totalRpcJson;
+
+    Json::Value jqobj(Json::objectValue);
+    // totalJq represents all jobs. All enqueued, started, finished, etc.
+    Jq totalJq("total");
+    for (auto const& proc : jq_)
+    {
+        Json::Value j(Json::objectValue);
+        {
+            std::unique_lock<std::mutex> lock(proc.second.mut);
+            auto const tmp = proc.second.sync;
+            lock.unlock();
+            j[jss::queued] = std::to_string(tmp.queued);
```

Comment:
Consider a similar optimization to the possible RPC optimization.

Reply:
fixed